### ISSUE TITLE
Support for `SELECT DISTINCT ON`

### ISSUE THREAD

**Issue by duairc** [tomjaguarpaw/haskell-opaleye#244]
At the moment is there any support for PostgreSQL's [DISTINCT ON](https://www.postgresql.org/docs/current/static/sql-select.html#SQL-DISTINCT) feature?

If not (I can't see it anywhere), what would be the best work-around in the mean time?

**Comment by tomjaguarpaw** [comment]
Currently there's no support for `DISTINCT ON`.

Could you aggregate into arrays and then take the first element of each array?

**Comment by fieldstrength** [comment]
I have also come to a situation where this `distinct on` support would be very useful to me.

I'm curious @tomjaguarpaw if you have any thoughts on an implementation plan that would be acceptable? This is valuable enough that I might probably give it a go, if its not too heavy a lift.

Some observations and questions:

* According to the [Postgres documentation](https://www.postgresql.org/docs/current/static/sql-select.html#SQL-DISTINCT) the allowed expressions are similar to those of `ORDER BY`, so the implementation might look similar. Like orders, we could use the nice semigroup interface, where different field projections can be appended together. So I'm thinking a reasonable interface would be something like

```haskell
data DistinctOn a 
-- = DistinctOn (a -> NonEmpty PrimExpr)

instance Semigroup DistinctOn
instance Contravariant DistinctOn

distinguish :: (a -> Column b) -> DistinctOn a

distinctOn :: DistinctOn a -> Query a -> Query a
```
* What should be done about our `distinct` function, currently implemented as a `groupBy` aggregation? While it might be nice to handle it together with this functionality, or actually make the implementation match the name, it seems like a better first step to consider this as an independent feature.

* There is a potential gotcha in that if one simultaneously uses `DISTINCT ON` with `ORDER BY`, then any columns appearing in the former have to appear before (to the left) of any other columns in the latter. While this was at first a worry, I believe this pitfall is sidestepped if we ensure that this new feature is kept at a different level of nesting as the ordering. If I have understood correctly, this is what the `simpleQueryArr` function does, is that correct?

**Comment by tomjaguarpaw** [comment]
Hi @fieldstrength.  Yes, please give this a go!  I'd probably just go for

```haskell
data DistinctOnPP a b = ...

instance ProductProfunctor DistinctOnPP where
    ....

distinctOn :: Default DistinctOnPP b b => (a -> b) -> Query a -> Query a
```

I think it's probably sufficient for `type DistinctOnPP = Unpackspec`, actually.

> What should be done about our distinct function

Just keep it as it is.  It's the interface that's important.  It doesn't matter if similar functionality is implemented in different ways.  We can merge them later at our leisure.

> columns appearing in the former have to appear before (to the left) of any other columns in the latter

I believe the emitted SQL should contain explicitly nested `SELECT`s, so this won't be a problem.

**Comment by duairc** [comment]
Fixed by #392
